usr/src/uts/common/io/scsi/adapters/smartpqi/smartpqi_sis.c

```diff
@@ -0,0 +1,251 @@
+#define	SIS_SMARTARRAY_FEATURES_SUPPORTED	0x2
+#define	SIS_PQI_MODE_SUPPORTED			0x4
+#define	SIS_REQUIRED_EXTENDED_PROPERTIES	\
+	(SIS_SMARTARRAY_FEATURES_SUPPORTED | SIS_PQI_MODE_SUPPORTED)
+
+#pragma pack(1)
+/* used for passing command parameters/results when issuing SIS commands */
+typedef struct sis_sync_cmd_params {
+	uint32_t	mailbox[6];	/* mailboxes 0-5 */
+} sis_sync_cmd_params_t;
+
+#define	SIS_BASE_STRUCT_REVISION	9
+
+typedef struct sis_base_struct {
+	uint32_t	sb_revision;
+	uint32_t	sb_flags;
+	uint32_t	sb_error_buffer_paddr_low;
+	uint32_t	sb_error_buffer_paddr_high;
+	uint32_t	sb_error_elements_len;
+	uint32_t	sb_error_elements_num;
+} sis_base_struct_t;
+#pragma pack()
+
+/* ---- Forward declaration for support functions ---- */
+static boolean_t sis_send_sync_cmd(pqi_state_t s, uint32_t cmd,
+    sis_sync_cmd_params_t *params);
+
+uint32_t
+sis_read_scratch(pqi_state_t s)
+{
+	return (G32(s, sis_driver_scratch));
+}
+
+void
+sis_write_scratch(pqi_state_t s, int mode)
+{
+	S32(s, sis_driver_scratch, mode);
+}
+
+boolean_t
+sis_reenable_mode(pqi_state_t s)
+{
+	int		loop_count;
+	uint32_t	doorbell;
+
+	S32(s, sis_host_to_ctrl_doorbell, SIS_REENABLE_SIS_MODE);
+
+	for (loop_count = 0; loop_count < 1000; loop_count++) {
+		doorbell = G32(s, sis_ctrl_to_host_doorbell);
+		if ((doorbell & SIS_REENABLE_SIS_MODE) == 0) {
+			return (B_TRUE);
+		}
+		drv_usecwait(1000);	/* ---- Wait 1ms ---- */
+	}
+	return (B_FALSE);
+}
+
+boolean_t
+sis_wait_for_ctrl_ready(pqi_state_t s)
+{
+	int		loop_count;
+	uint32_t	status;
+
+	for (loop_count = 0; loop_count < 1000; loop_count++) {
+		status = G32(s, sis_firmware_status);
+		if (status & SIS_CTRL_KERNEL_PANIC)
+			return (B_FALSE);
+		if (status & SIS_CTRL_KERNEL_UP)
+			return (B_TRUE);
+		drv_usecwait(1000);	 /* ---- Wait 1ms ---- */
+	}
+	return (B_FALSE);
+}
+
+/*
+ * sis_get_ctrl_props -- Verify we're talking to controller that speaks PQI
+ */
+boolean_t
+sis_get_ctrl_props(pqi_state_t s)
+{
+	sis_sync_cmd_params_t	p;
+	uint32_t		property;
+	uint32_t		extended_property;
+
+	(void) memset(&p, 0, sizeof (p));
+	if (sis_send_sync_cmd(s, SIS_CMD_GET_ADAPTER_PROPERTIES, &p) == B_FALSE)
+		return (B_FALSE);
+
+	property = p.mailbox[1];
+	if (!(property & SIS_EXTENDED_PROPERTIES_SUPPORTED))
+		return (B_FALSE);
+
+	extended_property = p.mailbox[4];
+	if ((extended_property & SIS_REQUIRED_EXTENDED_PROPERTIES) !=
+	    SIS_REQUIRED_EXTENDED_PROPERTIES)
+		return (B_FALSE);
+
+	return (B_TRUE);
+}
+
+boolean_t
+sis_get_pqi_capabilities(pqi_state_t s)
+{
+	sis_sync_cmd_params_t	p;
+
+	(void) memset(&p, 0, sizeof (p));
+	if (sis_send_sync_cmd(s, SIS_CMD_GET_PQI_CAPABILITIES, &p) == B_FALSE)
+		return (B_FALSE);
+
+	s->s_max_sg_entries = p.mailbox[1];
+	s->s_max_xfer_size = p.mailbox[2];
+	s->s_max_outstanding_requests = p.mailbox[3];
+	s->s_config_table_offset = p.mailbox[4];
+	s->s_config_table_len = p.mailbox[5];
+	return (B_TRUE);
+}
+
+#define	SIS_ROUNDUP(a, align, type) \
```

Comment:
Do we have one of these in sysmacros.h?  If we do, can you use it, if we don't can you give a name that more clearly differentiates it?